weaviate/embedded.py

```diff
@@ -188,6 +199,41 @@ def check_supported_platform() -> None:
                 f"you want this: https://github.com/weaviate/weaviate/issues/3315"
             )
 
+    def stop(self) -> None:
+        if self.process is not None:
+            try:
+                self.process.terminate()
+                self.process.wait()
+            except ProcessLookupError:
+                print(
+                    f"""Tried to stop embedded weaviate process {self.process.pid}. Process was not found. So not doing
+                    anything"""
+                )
+            self.process = None
+
+    def ensure_running(self) -> None:
+        if self.is_listening() is False:
+            print(
+                f"Embedded weaviate wasn't listening on ports http:{self.options.port} & grpc:{self.options.grpc_port}, so starting embedded weaviate again"
+            )
+            self.start()
+
+    @abstractmethod
+    def start(self) -> None:
+        raise NotImplementedError()
+
+
+class EmbeddedDB(_EmbeddedBase):
+    def is_listening(self) -> bool:
```

Comment:
this seems identical to the one in _EmbeddedBase?

Reply:
Ah nice catch. `is_listening` should be abstract in the base class since its implementation is different for the two v3 and v4 versions